examples/addr2line.rs

```diff
@@ -0,0 +1,198 @@
+extern crate gimli;
+extern crate getopts;
+
+use std::env;
+use std::str::FromStr;
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let mut opts = getopts::Options::new();
+    opts.optopt("e", "exe", "Set the input file name (default is a.out)", "<executable>");
+
+    let matches = match opts.parse(&args[1..]) {
+        Ok(m) => m,
+        Err(f) => { panic!(f.to_string()); }
+    };
+
+    let file_path = match matches.opt_str("e") {
+        Some(f) => f,
+        None => "a.out".to_string()
+    };
+
+    let file = obj::open(&file_path);
+    if obj::is_little_endian(&file) {
+        do_work::<gimli::LittleEndian>(&file, &matches);
+    } else {
+        do_work::<gimli::BigEndian>(&file, &matches);
+    }
+}
+
+fn do_work<Endian>(file: &obj::File, matches: &getopts::Matches)
+    where Endian: gimli::Endianity
+{
+    let addrs: Vec<u64> = matches.free.iter().map(|x| u64::from_str_radix(x,16).expect("Failed to parse")).collect();
+
+    let aranges = obj::get_section(file, ".debug_aranges").expect("Can't addr2line with no aranges");
+    let aranges = gimli::DebugAranges::<Endian>::new(aranges);
+    let mut aranges = aranges.aranges();
+
+    let mut dies: Vec<Option<gimli::DebugInfoOffset>> = Vec::new();
+    dies.resize(addrs.len(), None); // XXXkhuey better way to do this?
+    while let Some(arange) = aranges.next_arange().expect("Should parse arange OK") {
+        let start = arange.start();
+        let end = start + arange.len();
+
+        for (i, addr) in addrs.iter().enumerate() {
+            match (*addr >= start, *addr < end) {
+                (true, true) => {
+                    dies[i] = Some(arange.debug_info_offset());
+                },
+                _ => continue,
+            }
+        }
+    }
+
+    let debug_info = obj::get_section(file, ".debug_info").expect("Can't addr2line with no debug_info");
+    let debug_info = gimli::DebugInfo::<Endian>::new(debug_info);
+    let debug_abbrev = obj::get_section(&file, ".debug_abbrev")
+        .expect("Does not have .debug_abbrev section");
+    let debug_abbrev = gimli::DebugAbbrev::<Endian>::new(debug_abbrev);
+    let debug_line = obj::get_section(file, ".debug_line").expect("Can't addr2line with no debug_line");
+    let debug_line = gimli::DebugLine::<Endian>::new(&debug_line);
+
+    for (index, die) in dies.iter().enumerate() {
+        let addr = addrs[index];
+        match *die {
+            None => println!("Found nothing"),
+            Some(d) => match debug_info.header_from_offset(d) {
+                Err(_) => continue,
+                Ok(h) => {
```

Comment:
Ah, I missed that. You could do

```rust
let dies = dies.map(|opt_d| opt_d.map(|d| debug_info.header_from_offset(d)));
```

before the iteration. Whether or not that is an improvement... /me shrugs